Fixed clearCookies function for WkWebView

davidestrati opened the pull request:
This is using the proper method to clear the cookies when you are using WkWebview and have iOS 11 or above.
Plus it will maintain compatibility for lower version of iOS using the legacy method.

Commit:
Update plugin to work with Cordova >7.0.0

Details here:
https://cordova.apache.org/plugins/?q=cookie&platforms=cordova-android%2Ccordova-ios&sortBy=Recently%20Updated

Commit:
Update plugin.xml

edwardgaoyb committed:
Merge pull request #1 from TheTostu/patch-1

Update plugin to work with Cordova >7.0.0

davidestrati committed:
fixed clearcookies for wkwebview

davidestrati committed:
Merge pull request #1 from davidestrati/fix/clear-cookies-wkwebview

fixed clearcookies for wkwebview

davidestrati committed:
manifest fix

davidestrati committed:
manifest real

davidestrati committed:
manifest real

KevinKelchen commented on src/ios/CDVCookieMaster.m:
```diff
@@ -71,11 +71,22 @@ - (void)setCookieValue:(CDVInvokedUrlCommand*)command
+
     [[NSUserDefaults standardUserDefaults] synchronize];
 
     CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
```
Similar to how WKWebView support was added for `setCookieValue`, we should be wrapping the iOS 11 check in a check to see if the app is running in WKWebView. This could otherwise create issues for apps running in UIWebView. I know that Apple will no longer allow updates of apps using UIWebView beginning in December 2020, but for now we should continue to support it. 🙂 Here's the check I'm referring to:
https://github.com/Cartegraph/cordova-cookie-master/commit/9f7fb02ecbd4373b23b527e1cb19acee04c20741#diff-4a1ce2d830a973505b873c255085e7b6R61

Below are my suggested changes for this PR. Besides the addition of the WKWebView check and handling the failure case, I moved the `synchronize` call into the UIWebView branch of logic because I don't think it's necessary going forward with WKWebView. See here: https://developer.apple.com/documentation/foundation/nsuserdefaults/1414005-synchronize?language=objc . I also adjusted some code indentation.

Thanks a lot for contributing the PR! Let me know your thoughts. 🙂

```suggestion
    CDVPluginResult* pluginResult = nil;
    
    if ([self.webView isKindOfClass:[WKWebView class]]) {
        if (@available(iOS 11.0, *)) {
            NSSet *websiteDataTypes = [NSSet setWithArray:@[WKWebsiteDataTypeCookies]];
            NSDate *dateFrom = [NSDate dateWithTimeIntervalSince1970:0];
            [[WKWebsiteDataStore defaultDataStore] removeDataOfTypes:websiteDataTypes
                                                    modifiedSince:dateFrom
                                                    completionHandler:^{
                                                        NSLog(@"Cookies Cleared");
                                                    }];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"WKWebView requires iOS 11+ in order to clear cookies"];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];

            return;
        }
    } else {
        NSHTTPCookie *cookie;
        NSHTTPCookieStorage *storage = [NSHTTPCookieStorage sharedHTTPCookieStorage];
        for (cookie in [storage cookies]) {
            [storage deleteCookie:cookie];
        }
           
        [[NSUserDefaults standardUserDefaults] synchronize];
    }
    
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
```

KevinKelchen reviewed (changes requested):
Sorry it's taken so long to get around to this! Thanks for being so patient. 🙂

Please see the comment I've made and the suggested change.

Also, FYI that yesterday we released an update that exposes a new method which allows users to clear a single cookie by url  + cookie name. It sets an explicitly invalid cookie to effectively clear the valid cookie because that seemed like the most straightforward implementation. We added this feature because we didn't want to clear all cookies in our app (including third-party cookies) by using the existing `clearCookies` method. This new method allows users to target a specific cookie.

KevinKelchen commented:
Also, I'm not sure what was behind all the intermediate commits to the `plugin.xml` file regarding the `AndroidManifest.xml` path. Regarding the commit message to make it work with Cordova versions greater than 7, this plugin already works with Cordova 9 (which is what our app uses). In the end it looks like no changes were ultimately made though.

If we merge this PR we will likely squash those commits to keep the repo history clean.

Commit:
accepted suggestion and improved clear cookie method

davidestrati committed:
Merge branch 'master' into master

davidestrati commented:
Hi Kevin,

Sorry for the long time that this took up.

I implemented your suggestions and modified the `setCookie` method as long as the `clearCookie` one.
I do believe that this method could still be useful in some rare occasions or in applications that do not take advantage of cookie retention and just need to delete them.

As regards the multiple commits for cordova 9 compatibility, it was clearly my mistake, so you can ignore and squash them if you like.

Thank you for all your suggestions.

Commit:
updated with correct version for clear cookie function availability in wkwebview

Commit:
Merge branch 'master' of https://github.com/davidestrati/cordova-cookie-master

davidestrati committed:
fixed typo in string

davidestrati reviewed (commented):
Improved with suggestions

KevinKelchen commented:
There appear to be some discrepancies from the suggested changes.

1. `pluginResult` is not declared at the top of the method.
2. Using iOS 9 as the version check instead of being consistent with `setCookie`'s iOS 11 check. What was the reason for this?
3. Some indentation is off in the UIWebView-related `else` statement.
4. The `synchronize` call is not moved into the UIWebView-related `else` statement.

I would have thought we could have accepted the entire suggested change in the PR which would have prevented these issues. Or at least a blanket copy paste. Were there other issues involved with the suggested changes?

I'd be happy to open a new PR to make those changes if that is helpful. 🙂

Commit:
improved with suggested changes

davidestrati commented:
Hi Kevin.

I improved the PR with your suggested changes.

The reason why I used iOS 9 check instead of 11, was that the method that it is using requires iOS 9 or greater. However I replaced it with iOS 11 for consistency.

As regards the original suggested changes, I tried to merge them with my modifications in order to make `clearCookies` method work correctly.

If you believe that this PR is not relevant or useful anymore, feel free to close it. I just wanted to help! 😅

Regards
